### ISSUE TITLE
block-scoped-var issue with modules: true

### ISSUE THREAD

**Issue by hzoo** [eslint/eslint#2625]
`warning  "wrapUserProvidedKey" used outside of binding context  block-scoped-var`

file.js
```js
function getComponentKey(component, index) {
  if (component && component.key != null) {
    return wrapUserProvidedKey(component.key); // here
  }
  return index.toString(36);
}

function wrapUserProvidedKey(key) {
  return '$' + key;
}
```

.eslintrc
```yaml
ecmaFeatures:
  modules: true

rules:
  block-scoped-var: 1
```

Tested on master and latest ESLint.

Ref: https://github.com/facebook/react/pull/3968

I just picked a single file to isolate the issue.

Actual lines: https://github.com/facebook/react/blob/52752446760dee0bc7232b4146f5a309ac57f065/src/shared/utils/traverseAllChildren.js#L51-L82

**Comment by hzoo** [comment]
Oh after some searching this looks to be the same as https://github.com/eslint/eslint/issues/2291 actually.

**Comment by michaelficarra** [comment]
Yep, you give up function declaration hoisting when you enable that rule.

**Comment by hzoo** [comment]
Ok cool will close.
